### PR TITLE
[SLP] Enable reordering for non-power-of-two vectors

### PR DESCRIPTION
**Pull request by preames**
This change tries to enable vector reordering during vectorization for non-power-of-two vectors.  Specifically, my goal is to be able to vectorize reductions whose operands appear in other than identity order.  (i.e. a[1] + a[0] + a[2]).  Our standard pass pipeline, Reassociation effectively canonicalizes towards this form. So for reduction vectorization to be wildly applicable, we need this feature.

This change enables the use of a non-empty ReorderIndices structure - which is effectively required for out of order loads or gathers - while leaving the ReuseShuffleIndices mechanism unused and disabled. If I've understood the code structure, the former is used when describing implicit shuffles required by the vectorization strategy (i.e. loading elements 0,1,3,2 in the order 0,1,2,3 and then shuffling later), while the later is used when trying to optimize explode/buildvectors (called gathers in this code).

I auditted all the code enabled by this change, but can't claim to deeply understand most of it.  I added a couple of bailouts in places which appeared to be difficult to audit and optional optimizations. I've tried to do so in the least risky way I can, but am not completely confident in this change.  Careful review appreciated.

### CODE REVIEW

**Comment by preames** [comment]
ping

**Comment by alexey-bataev** [comment]
Please rebase

**Comment by preames** [comment]
> Please rebase

I have asked for a revert of the change causing the bad merge, please see: https://github.com/llvm/llvm-project/pull/106449

**Comment by preames** [comment]
Rebased.

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 1]
```diff
@@ -5424,11 +5422,13 @@ BoUpSLP::getReorderingData(const TreeEntry &TE, bool TopToBottom) {
...
     if (isSplat(TE.Scalars))
       return std::nullopt;
-    if (TE.Scalars.size() >= 4)
+    if (TE.Scalars.size() >= 3)
       if (std::optional<OrdersType> Order = findPartiallyOrderedLoads(TE))
         return Order;
-    if (std::optional<OrdersType> CurrentOrder = findReusedOrderedScalars(TE))
-      return CurrentOrder;
+
+    if (!TE.isNonPowOf2Vec())
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:5429]: Add a FIXME to remove this check later

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 2]
```diff
@@ -3388,6 +3388,8 @@ class BoUpSLP {
     TreeEntry *Last = VectorizableTree.back().get();
     Last->Idx = VectorizableTree.size() - 1;
     Last->State = EntryState;
+    assert((has_single_bit(VL.size()) || ReuseShuffleIndices.empty()) &&
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:3391]: Add a FIXME to remove this later

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 3]
```diff
@@ -4824,6 +4817,12 @@ BoUpSLP::LoadsState BoUpSLP::canVectorizeLoads(
   // representation is better than just gather.
   auto CheckForShuffledLoads = [&, &TTI = *TTI](Align CommonAlignment,
                                                 bool ProfitableGatherPointers) {
+    // The following code has not been updated for non-power-of-2 vectors.
+    // The splitting logic here does not cover the original vector if the
+    // vector factor is not a power of two.
+    if (!has_single_bit(VL.size()))
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:4823]: Add a FIXME

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 4]
```diff
@@ -5195,13 +5194,12 @@ static bool areTwoInsertFromSameBuildVector(
...
 BoUpSLP::getReorderingData(const TreeEntry &TE, bool TopToBottom) {
-  // FIXME: Vectorizing is not supported yet for non-power-of-2 ops.
-  if (TE.isNonPowOf2Vec())
-    return std::nullopt;
-
   // No need to reorder if need to shuffle reuses, still need to shuffle the
   // node.
   if (!TE.ReuseShuffleIndices.empty()) {
+    assert(!TE.isNonPowOf2Vec() &&
+           "Reshuffling scalars not yet supported for nodes with padding");
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:5201]: Add a FIXME

**Review by alexey-bataev — APPROVED** [review]

**Comment by preames** [comment]
> When pulling this down through internal merge, I found a problem (crash). I'm reducing now, and will either post a fix short if this turns out to be easy, or revert.

Fixed in 247d3ea.

**Pull request merged**